### PR TITLE
Introduce structs to high-level P3 interface

### LINKED ISSUE #465: Reorganize top-level P3 interface to use structs, not individual variables.
Here's the conversation surrounding this task:

https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/1594786870/Pass+Structs+Not+Vomit?focusedCommentId=1621262995#comment-1621262995

Briefly, we're introducing a number of structs to gather input/output data for use with `p3_main`:

+ `P3PrognosticState: qc, nc, qr, nr, th, qv, qitot, qirim, nitot, birim`

+ `P3DiagnosticInputs: ncnuc, naai, qc_relvar, icldm, lcldm, rcldm, pres, dzq, pdel, exner`

+ `P3DiagnosticOutputs: mu_c, lamc, cmeiout, prt_liq, prt_sol, diag_effc, diag_effi, diag_rhoi, prain, nevapr, prer_evap, rflx, sflx`

+ `P3Infrastructure: dt, it, its, ite, kts, kte, log_predictNc, col_location`

+ `P3HistoryOnly: iq_ice_exchange, vap_liq_exchange, vap_ice_exchange`

This task consists of the following steps:

1. Clean up the p3_main function interface (and its callers) to remove variables we don’t actually use and to order the variables according to their grouping into the above structs.

2. Clean up the boilerplate code to expect structs rather than individual variables.

3. Prepend the appropriate “<struct_name>.” in front of all non-local variables within the actual `p3_main` code.

4. Modify `p3_main_unit_tests.cpp` and `p3_run_and_cmp.cpp` (the only functions which call `p3_main`) to pass structs.

### PR DESCRIPTION
**Pull request by jeff-cohere**
Here's the refactor that changes P3 to use structs. I have to say: it takes a lot of concentration to work in this environment. Maybe we could make it easier with more comments or some more tooling to help us perform memory-related operations. As it is, it can be fairly challenging to alter an interface once we've ported it.

+ P3 structs have been introduced in p3_functions.hpp as defined in #465 .
+ `p3_main` now accepts these structs as arguments. We still have `ni` and `nk` as separate integer parameters.
+ The Fortran bridge functions still all use separate variables, including `p3_main_f`, which Fortran calls to access our C++ implementation.
+ I've removed the `diag_ze`,  `diag_vmi`, `diag_di`, `pratot`, and `prctot` parameters from `p3_main`, both on the C++ and Fortran side. However, the lower-level functions still use these.
+ I've removed the `log_` prefix from boolean variables, since (a) it's no longer 1977 and variable names don't need to indicate types, and (b) these variables are related neither to logarithms nor logging. 

The `atmosphere_microphysics.*` code uses the F90 interface, so I haven't made any changes to it.

Closes #465

### CODE REVIEW

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 1]
```diff
@@ -1149,6 +1143,14 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th,qv,dt,qitot,qirim,nitot,birim,   &
     !----- Local variables and parameters:  -------------------------------------------------!
     !
 
+    ! These outputs are no longer provided by p3_main.
```
Comment (jgfouca) [inline components/cam/src/physics/cam/micro_p3.F90:1146]: Why not?

File: components/scream/src/physics/p3/p3_functions.hpp [hunk 2]
```diff
@@ -753,7 +862,7 @@ struct Functions
     const uview_1d<Spack>& birim_incld,
     const uview_1d<Spack>& omu_c,
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions.hpp:863]: `omu_c` should be `mu_c` (just like you fixed the lamc name).

File: components/scream/src/physics/p3/p3_functions.hpp [hunk 3]
```diff
@@ -92,6 +92,115 @@ struct Functions
 
   using Workspace = typename WorkspaceManager<Spack, Device>::Workspace;
 
+  // This struct stores prognostic variables evolved by P3.
+  struct P3PrognosticState {
```
Comment (bartgol) [inline components/scream/src/physics/p3/p3_functions.hpp:96]: How about putting these types in a separate header, say, `p3_data_structures.hpp`? It might help compartmentalize the code a bit more. There's a clear dependency: interfaces require data structures, but not viceversa.

File: components/scream/src/physics/p3/p3_main_impl.hpp [hunk 4]
```diff
@@ -963,121 +919,123 @@ ::p3_main(
...
-    const auto onevapr           = util::subview(nevapr, i);
-    const auto oprer_evap        = util::subview(prer_evap, i);
-    const auto orflx             = util::subview(rflx, i);
-    const auto osflx             = util::subview(sflx, i);
-    const auto opratot           = util::subview(pratot, i);
-    const auto oprctot           = util::subview(prctot, i);
-    const auto oliq_ice_exchange = util::subview(liq_ice_exchange, i);
-    const auto ovap_liq_exchange = util::subview(vap_liq_exchange, i);
-    const auto ovap_ice_exchange = util::subview(vap_ice_exchange, i);
+    const auto opres             = util::subview(diagnostic_inputs.pres, i);
```
Comment (bartgol) [inline components/scream/src/physics/p3/p3_main_impl.hpp:949]: Out of curiosity, what does the "o" prefix stand for? I can't think of any o-word that express locality to a single column. My actual first guess was 'output', but since all the subviews have it, it's not that.

In case you are not a fan of all those oxyz names, notice that, since the inputs are now inside a struct, the names _without_ the prefix are again available for use.

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 1 above)
Comment (AaronDonahue) [reply components/cam/src/physics/cam/micro_p3.F90:1146]: They were added in the first round of p3 development because they are present in MG2 and we were using that as a template.  Subsequent code clean-up tasks have removed their use in the post-processing so they are no longer needed explicitly from P3.

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 5]
```diff
@@ -1105,14 +1105,10 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th,qv,dt,qitot,qirim,nitot,birim,   &
...
     real(rtype), intent(out),   dimension(its:ite)              :: prt_sol    ! precipitation rate, solid        m s-1
-    real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: diag_ze    ! equivalent reflectivity          dBZ
     real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: diag_effc  ! effective radius, cloud          m
     real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: diag_effi  ! effective radius, ice            m
-    real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: diag_vmi   ! mass-weighted fall speed of ice  m s-1
-    real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: diag_di    ! mean diameter of ice             m
     real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: diag_rhoi  ! bulk density of ice              kg m-3
     real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: mu_c       ! Size distribution shape parameter for radiation
-    real(rtype), intent(out),   dimension(its:ite,kts:kte)      :: lamc       ! Size distribution slope parameter for radiation
```
Comment (AaronDonahue) [inline components/cam/src/physics/cam/micro_p3.F90:1115]: I must have forgotten to add lamc to one of the structures on Confluence, but it is still in use and belongs with the same structure as mu_c.

File: components/scream/src/physics/p3/p3_functions.hpp (hunk 3 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_functions.hpp:96]: Yeah, I guess I stuck things in here to get the right typing information for the views in the struct from the template parameters of the `Functions` "namespace" struct. Further, these structs are only ever used in conjunction with `p3_main`--I don't think we'd ever use them without calling that specific function. So it seemed expedient to place them alongside it.

That said, I don't have a strong opinion about where these structs live. How do others feel about this? Should we punt on this for now and discuss this in the context of Luca's proposed restructuring (#478)?

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 5 above)
Comment (jeff-cohere) [reply components/cam/src/physics/cam/micro_p3.F90:1115]: Okay, I'll wire it back up.

File: components/scream/src/physics/p3/p3_main_impl.hpp (hunk 4 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_main_impl.hpp:949]: @jgfouca , can you comment on this? I've just been treading lightly here, and it'd be interesting to hear what you intended with this prefix.

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 5 above)
Comment (jeff-cohere) [reply components/cam/src/physics/cam/micro_p3.F90:1115]: What about `lam_r`? We also have `mu_r` in `P3DiagnosticsOutputs.

File: components/scream/src/physics/p3/p3_functions.hpp (hunk 3 above)
Comment (bartgol) [reply components/scream/src/physics/p3/p3_functions.hpp:96]: Yes, they will have a limited sets of customers: some high level p3 function, and atmosphere_microphysics.cpp. I just think it makes it easy to track down if they are isolated to a single file. But again, not a big deal (by any means), given the limited scope of the structs. So unless someone else agrees that putting them in a separate headers improves readability/maintainability, then they can stay where they are.

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 5 above)
Comment (AaronDonahue) [reply components/cam/src/physics/cam/micro_p3.F90:1115]: maybe that was the typo, my quick check is that `mu_r` wasn't a typical output of P3, so I can't imagine it is being used.  So maybe I wrote lamc and mu_r when it should be lamc and mu_c.

File: components/scream/src/physics/p3/p3_functions.hpp (hunk 3 above)
Comment (bartgol) [reply components/scream/src/physics/p3/p3_functions.hpp:96]: And yes, maybe a discussion about this in the issue you linked might be more appropriate.

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 6]
```diff
@@ -3510,20 +3521,15 @@ void p3_main_f(
...
-    diag_ze_d          (temp_d[counter++]),
     diag_effc_d        (temp_d[counter++]),
     diag_effi_d        (temp_d[counter++]),
-    diag_vmi_d         (temp_d[counter++]),
-    diag_di_d          (temp_d[counter++]),
     diag_rhoi_d        (temp_d[counter++]),
     mu_c_d             (temp_d[counter++]),
-    lamc_d             (temp_d[counter++]),
+    mu_r_d             (temp_d[counter++]),
     cmeiout_d          (temp_d[counter++]),
```
Comment (AaronDonahue) [inline components/scream/src/physics/p3/p3_functions_f90.cpp:3529]: does this mean we've removed lamc and replaced with mu_r?

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 7]
```diff
@@ -3559,22 +3576,25 @@ void p3_main_f(
   // Sync back to host
   Kokkos::Array<view_2d, P3MainData::NUM_ARRAYS - 10> inout_views = {
     qc_d, nc_d, qr_d, nr_d, qitot_d, qirim_d, nitot_d, birim_d, qv_d, th_d,
-    diag_ze_d, diag_effc_d, diag_effi_d, diag_vmi_d, diag_di_d, diag_rhoi_d, mu_c_d, lamc_d, cmeiout_d, prain_d, nevapr_d, prer_evap_d, pratot_d, prctot_d, liq_ice_exchange_d, vap_liq_exchange_d, vap_ice_exchange_d, rflx_d, sflx_d, prt_liq_temp_d, prt_sol_temp_d
```
Comment (AaronDonahue) [inline components/scream/src/physics/p3/p3_functions_f90.cpp:3562]: same here with lamc -> mu_r

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 6 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:3529]: Yes. I implemented the structs as described in #465 . Can you edit the description of that issue and fix the definitions of the structs to your liking?

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 8]
```diff
@@ -3559,22 +3576,25 @@ void p3_main_f(
...
-  dim1_sizes_out[29] = 1; dim2_sizes_out[29] = ni; // prt_liq
-  dim1_sizes_out[30] = 1; dim2_sizes_out[30] = ni; // prt_sol
+  dim2_sizes_out[21] = nk+1; // rflx
+  dim2_sizes_out[22] = nk+1; // sflx
+  dim1_sizes_out[23] = 1; dim2_sizes_out[23] = ni; // prt_liq
+  dim1_sizes_out[24] = 1; dim2_sizes_out[24] = ni; // prt_sol
 
   pack::device_to_host({
-      qc, nc, qr, nr, qitot, qirim, nitot, birim, qv, th,
-      diag_ze, diag_effc, diag_effi, diag_vmi, diag_di, diag_rhoi, mu_c, lamc, cmeiout, prain, nevapr, prer_evap, pratot, prctot, liq_ice_exchange, vap_liq_exchange, vap_ice_exchange, rflx, sflx, prt_liq, prt_sol
```
Comment (AaronDonahue) [inline components/scream/src/physics/p3/p3_functions_f90.cpp:3577]: same here with lamc -> mu_r (just marking everywhere I notice it just in case you need to change it and would appreciate knowing where a few might be :-)

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 6 above)
Comment (AaronDonahue) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:3529]: absolutely.  I'll do that now

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 8 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:3577]: Thanks! Appreciated.

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 6 above)
Comment (AaronDonahue) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:3529]: Okay, all done.  I fixed Confluence too.  Not sure how those got switched, I can't even pretend that `mu_r` and `lamc` are next to each other on the keyboard...  I would blame autocorrect, but that's probably a bridge to far.
Sorry to make sure change all these just to change them back.  Glad we caught it though.

Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:3529]: Thanks! And no worries--this is exactly what PRs and code reviews are for.

File: components/scream/src/physics/p3/p3_main_impl.hpp [hunk 9]
```diff
@@ -73,7 +73,7 @@ void Functions<S,D>
 ::p3_main_part1(
   const MemberType& team,
   const Int& nk,
-  const bool& log_predictNc,
+  const bool& predictNc,
```
Comment (PeterCaldwell) [inline components/scream/src/physics/p3/p3_main_impl.hpp:76]: Out of curiosity, why did you rename log_X to X. Don't get me wrong - I hated log_X because I always thought we were taking the log of something - but it just seems orthogonal to the point of this PR and I thought @xyuan was going to switch these logicals to is_X in an upcoming PR.

File: components/scream/src/physics/p3/p3_main_impl.hpp (hunk 4 above)
Comment (PeterCaldwell) [reply components/scream/src/physics/p3/p3_main_impl.hpp:949]: I think the goal was just to be able to have 2 different variables with basically the same name. Maybe o stands for One-column or lOcal? In any case, using the fact that inputs are in a struct to get rid of those mysterious "o"s sounds like a good idea to me. But we should hear @jgfouca 's opinion before following my brash judgement.

File: components/scream/src/physics/p3/p3_main_impl.hpp (hunk 9 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_main_impl.hpp:76]: Coincidentally, I added a bullet to the description above about this. I guess I couldn't help myself since I was already dealing with these variables in the struct. I can change them back if you think it complicates the aforementioned PR. (though I think simply replacing `log_` with `is_` might produce some awkward pseudo-English [though software devs don't seem to mind this particular outcome]).

Comment (bartgol) [reply components/scream/src/physics/p3/p3_main_impl.hpp:76]: +1 for the doubt on whether `is_predictNc` is better. Imho the "is_" prefix works if the following word is an adjective (or a noun at best), not a verb. For verbs, no prefix is better: `computeA` is better than `is_computeA`.

Comment (PeterCaldwell) [reply components/scream/src/physics/p3/p3_main_impl.hpp:76]: Yeah, I noticed that is_predictNC sounds horrible. I'm caught between wanting to unambiguously signal that a variable is a logical and not wanting to sound like a russian troll. Maybe do_predictNC? In any case, I think your change is fine - I mostly just wanted to tag Xingqiu so he's aware of the change when he looks for these logical variables.

Comment (bartgol) [reply components/scream/src/physics/p3/p3_main_impl.hpp:76]: I gues "do_blah" is ok. I don't have a strong opiion between options like "computeA" or "do_computeA". Whatever you guys decide there, works for me.

Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_main_impl.hpp:76]: To me, the verb is enough of an indicator, but I don't feel strongly enough to oppose this idea. I do kind of dislike the mixing of snake case and camel case, but that's a whole other can of worms. :-)

**Comment by PeterCaldwell** [comment]
Thanks for your heroic work on this structs PR, Jeff! We should have more of a post-mortem during the physics telecon on Monday, but start thinking about what we learn from this PR now. Some initial thoughts:

- I was hoping switching to structs would reduce the amount of boilerplate we have. Instead it looks like I just forced you to traverse that boilerplate in its entirety. Would having F90 also use these structs have simplified all that bridging stuff? What can we do now to make using structs like this in SHOC easier? Do we pack up the F90 vars now in a cleanup PR? Or do we just assert that we should switch to structs as part of our top-level function porting?
- I was originally thinking we should also use structs for p3_main_part1 through part 3. But maybe we should wait on that until we totally abandon the F90 and its bridging to make life easier?

File: components/scream/src/physics/p3/p3_main_impl.hpp (hunk 9 above)
Comment (PeterCaldwell) [reply components/scream/src/physics/p3/p3_main_impl.hpp:76]: Yeah, camel-snake is an abomination. Let's not do anything here and instead fight this battle in Xingqiu's PR. I'm cool with either way as well.

File: components/scream/src/physics/p3/p3_main_impl.hpp (hunk 4 above)
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_main_impl.hpp:949]: @bartgol , the "o" prefix means it's a single column subview of a 2D view

**Comment by jgfouca** [comment]
@PeterCaldwell , one problem with this PR is that it combines the Struct work with changing the parameters to a number of functions, which caused a lot of churn across all the bridges and tests etc. I think the PR would have been a lot smaller if it only did the struct change.

**Comment by jeff-cohere** [comment]
I agree with Jim on this one.

Also, I think the earlier we use structs and the more pervasively we adopt them throughout, the more benefits we'll gain from them. Though I suspect there's going to be a non-trivial mapping between the C++/Kokkos structs and the Fortran ones, we can probably discuss how to minimize such things.

**Comment by bartgol** [comment]
@PeterCaldwell even if the amount of code does not change much, I think there is another advantage in having structs (which I'm not sure we discussed in the confluence page or during telecons): when looking at the code, one now can clearly see how a field fits in P3's inputs or outputs. E.g., is this a prognostic or diagnostic variable? Is it computed just for history output or is it actually _needed_ in the prognostic state?

**Comment by jeff-cohere** [comment]
After chatting with @bartgol , I'm inclined to leave the `atmosphere_microphysics.*` changes for a later PR, since that machinery currently uses the F90 interface. That means, pending any further requests for changes, this PR's ready to go! I'll take the [WIP] tag off the title and see what the AT thinks. In the meantime, please give this one more look and let me know if you see anything that worries you that we haven't discussed already.

**Comment by PeterCaldwell** [comment]
What are the "the atmosphere_microphysics.* changes"?

**Comment by jeff-cohere** [comment]
> What are the "the atmosphere_microphysics.* changes"?

Sorry for not being more clear. These are changes to our atmospheric driver, which currently calls the P3 F90 implementation. We originally discussed the possibility of needing to change this code as part of this PR, but since I ended up preserving the F90 interface, we don't need to do anything yet. When the AD starts using the C++ version of P3, it will have to be changed to use our new P3 structs.

**Comment by PeterCaldwell** [comment]
Perfect. I agree we should wait on the AD side until its C++ interface is done.

**Comment by jeff-cohere** [comment]
Hmm. Looks like this PR isn't "B4B" after all, since it eliminates some outputs from the baseline files. I generated baselines on my local machine and the tests passed. @jgfouca , @bartgol , and @PeterCaldwell , does this sound okay to you? If so, I'll remove the B4B tag and rerun the tests.

**Comment by PeterCaldwell** [comment]
Yeah, I've dealt with similar situations. I'm comfortable with your plan.

**Comment by bartgol** [comment]
Jeff, did you test on a gpu machine as well? Are we sure the non bfb changes are _only_ due to those missing output fields? If so, we can manually merge.

**Comment by jeff-cohere** [comment]
Good point--I haven't done that yet. I'll try it out on Cori.

**Comment by bartgol** [comment]
Well, it seems highly likely given that the only failing tests are `p3_run_and_cmp_[f90|cxx]`. But better to double check, if you can. Once you are sure of that, someone with admin powers can force-merge the PR.

**Comment by jeff-cohere** [comment]
Okay, `p3_run_and_cmp_{f90/cxx}` both check out on Cori. The other thing in this PR's favor is: if you check the actual errors, it's not a difference in the answers, it's a dimensional mismatch for fields, which supports the idea that the missing outputs in the new tests don't match up properly to the old baselines (which still contain these outputs we eliminated).

Okay to regenerate baselines and merge?
